File: func_parser/core/permissions.py
"""Permission checking for func_parser."""
from __future__ import annotations

from typing import Dict, List, Set

from .errors import PermissionDeniedError

__all__ = ["PermissionChecker"]


class PermissionChecker:
    """Checks user permissions for commands."""

    # role → set of allowed permissions
    ROLE_PERMISSIONS: Dict[str, Set[str]] = {
        "admin": {"*"},       # admin can do everything
        "user": set(),
        "moderator": {"moderate"},
    }

    def __init__(self) -> None:
        self._role_permissions: Dict[str, Set[str]] = {
            role: set(perms) for role, perms in self.ROLE_PERMISSIONS.items()
        }
        self._user_permissions: Dict[str, Set[str]] = {}  # user_id → extra permissions

    def add_role(self, role: str, permissions: List[str]) -> None:
        """Add permissions to a role (creates the role if absent)."""
        self._role_permissions.setdefault(role, set()).update(permissions)

    def grant(self, user_id: str, permission: str) -> None:
        """Grant an individual permission to a specific user."""
        self._user_permissions.setdefault(user_id, set()).add(permission)

    def revoke(self, user_id: str, permission: str) -> None:
        """Revoke an individual permission from a specific user."""
        self._user_permissions.get(user_id, set()).discard(permission)

    def check(
        self,
        user_roles: List[str],
        required_permissions: List[str],
        user_id: str = "",
    ) -> bool:
        """Return True if the user has **all** of *required_permissions*."""
        if not required_permissions:
            return True
        user_perms: Set[str] = set()
        for role in user_roles:
            role_perms = self._role_permissions.get(role, set())
            if "*" in role_perms:
                return True  # admin shortcut
            user_perms.update(role_perms)
        user_perms.update(self._user_permissions.get(user_id, set()))
        return all(p in user_perms for p in required_permissions)

    def require(
        self,
        user_roles: List[str],
        required_permissions: List[str],
        user_id: str = "",
    ) -> None:
        """Raise :class:`PermissionDeniedError` if any required permission is missing."""
        for perm in required_permissions:
            if not self.check(user_roles, [perm], user_id):
                raise PermissionDeniedError(perm, user_id)

File: func_parser_rs/src/permissions.rs
//! Permission checker for func_parser_rs.

use std::collections::{HashMap, HashSet};
use crate::errors::{FuncParserError, Result};

/// Checks user permissions for commands.
#[derive(Debug, Clone)]
pub struct PermissionChecker {
    role_permissions: HashMap<String, HashSet<String>>,
    user_permissions: HashMap<String, HashSet<String>>,
}

impl PermissionChecker {
    pub fn new() -> Self {
        let mut role_permissions = HashMap::new();
        role_permissions.insert("admin".to_string(), {
            let mut s = HashSet::new();
            s.insert("*".to_string());
            s
        });
        role_permissions.insert("user".to_string(), HashSet::new());
        role_permissions.insert("moderator".to_string(), {
            let mut s = HashSet::new();
            s.insert("moderate".to_string());
            s
        });
        Self {
            role_permissions,
            user_permissions: HashMap::new(),
        }
    }

    pub fn add_role(&mut self, role: impl Into<String>, permissions: Vec<String>) {
        self.role_permissions
            .entry(role.into())
            .or_default()
            .extend(permissions);
    }

    pub fn grant(&mut self, user_id: impl Into<String>, permission: impl Into<String>) {
        self.user_permissions
            .entry(user_id.into())
            .or_default()
            .insert(permission.into());
    }

    pub fn revoke(&mut self, user_id: &str, permission: &str) {
        if let Some(perms) = self.user_permissions.get_mut(user_id) {
            perms.remove(permission);
        }
    }

    pub fn check(&self, user_roles: &[String], required: &[String], user_id: &str) -> bool {
        if required.is_empty() {
            return true;
        }
        // Admin shortcut
        for role in user_roles {
            if let Some(perms) = self.role_permissions.get(role) {
                if perms.contains("*") {
                    return true;
                }
            }
        }
        // Collect all permissions for user
        let mut user_perms: HashSet<&str> = HashSet::new();
        for role in user_roles {
            if let Some(perms) = self.role_permissions.get(role) {
                user_perms.extend(perms.iter().map(|s| s.as_str()));
            }
        }
        if let Some(extra) = self.user_permissions.get(user_id) {
            user_perms.extend(extra.iter().map(|s| s.as_str()));
        }
        required.iter().all(|p| user_perms.contains(p.as_str()))
    }

    pub fn require(&self, user_roles: &[String], required: &[String], user_id: &str) -> Result<()> {
        for perm in required {
            if !self.check(user_roles, &[perm.clone()], user_id) {
                return Err(FuncParserError::PermissionDenied {
                    permission: perm.clone(),
                    user: user_id.to_string(),
                });
            }
        }
        Ok(())
    }
}

impl Default for PermissionChecker {
    fn default() -> Self {
        Self::new()
    }
}
